server/aesdsocket.c

```diff
@@ -12,14 +12,210 @@
+    char *temp = NULL;
+    off_t fsize;
+
+    while (1)
+    {
+        buflen = START_LEN;
+        pos = 0;
+        buffer = (char *)malloc(sizeof(char) * buflen);
+        if (buffer == NULL)
+        {
+            perror("malloc");
+            printf("buffer malloc failed\n");
+            thread_cleanup(buffer, sendBuf);
+            thread_data->m_thread_complete = true;
+            return thread_data;
+        }
+        memset(buffer, 0, buflen);
+        do
+        {
+            pthread_mutex_lock(thread_data->m_mutex_exit);
+            if (thread_data->m_exit_requested)
+            {
+                thread_cleanup(buffer, sendBuf);
+                thread_data->m_thread_complete = true;
+                pthread_mutex_unlock(thread_data->m_mutex_exit);
+                return thread_data;
+            }
+            pthread_mutex_unlock(thread_data->m_mutex_exit);
+
+            len_recv = recv(thread_data->m_conn_fd, buffer + pos, buflen - pos - 1, 0);
+            if (len_recv == -1)
+            {
+                if (errno == EAGAIN || errno == EWOULDBLOCK)
+                {
+                    continue;
+                }
+                perror("recv");
+                printf("Receive failed\n");
+                thread_cleanup(buffer, sendBuf);
+                thread_data->m_thread_complete = true;
+                return thread_data;
+            }
+            else if (len_recv == buflen - pos - 1)
+            {
+                buflen += START_LEN;
+                pos = buflen - START_LEN - 1;
+                temp = (char *)realloc(buffer, buflen);
+                if (temp == NULL)
+                {
+                    perror("realloc");
+                    printf("Realloc failed\n");
+                    thread_cleanup(buffer, sendBuf);
+                    thread_data->m_thread_complete = true;
+                    return thread_data;
+                }
+                buffer = temp;
+            }
+            else if (len_recv == 0)
+            {
+                thread_cleanup(buffer, sendBuf);
+                thread_data->m_thread_complete = true;
+                return thread_data;
+            }
+            else
+            {
+                pos += len_recv;
+            }
+            *(buffer + (buflen - 1)) = 0;
+        } while (strchr(buffer, '\n') == NULL);
+
+        pthread_mutex_lock(&mutex);
+        len_write = write(fd, buffer, strlen(buffer));
+        if (len_write == -1)
+        {
+            perror("write");
+            printf("Write failed\n");
+            thread_cleanup(buffer, sendBuf);
+            thread_data->m_thread_complete = true;
+            return thread_data;
+        }
+        else if (len_write != strlen(buffer))
+        {
+            perror("write");
+            printf("Partial write occurred\n");
+            thread_cleanup(buffer, sendBuf);
+            thread_data->m_thread_complete = true;
+            return thread_data;
```

Comment:
@zeog1488 missing mutex release in these error paths, better to unlock immediately after write